### ISSUE TITLE
copy_'s functionalized operator keeps copied into tensor live

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#102206]
### 🐛 Describe the bug

It's possible decomps fix this problem, but without decomps the graph is not as good as it could be:

```
import torch
from torch._functorch.aot_autograd import aot_export_joint_simple
    
def f(x, weight):
    def cb(grad_out):
        y = grad_out.sin()  # stand-in for collective result
        weight.data.copy_(y)
        # Better would be:
        #   weight.data = y
        return grad_out
    r = x @ weight
    r.register_hook(cb) 
    return (r,)

gm = aot_export_joint_simple(
    f,
    [torch.randn(2, 3, requires_grad=True), torch.randn(3, 4)],
    trace_joint=True,
)
gm.print_readable()
```

gives

```
class joint_helper(torch.nn.Module):
    def forward(self, primals, tangents):
        primals_1: f32[2, 3], primals_2: f32[3, 4], tangents_1: f32[2, 4], = fx_pytree.tree_flatten_spec([primals, tangents], self._in_spec)
        # No stacktrace found for following nodes
        mm: f32[2, 4] = torch.ops.aten.mm.default(primals_1, primals_2);  primals_1 = None
        sin: f32[2, 4] = torch.ops.aten.sin.default(tangents_1)
        copy: f32[3, 4] = torch.ops.aten.copy.default(primals_2, sin);  primals_2 = sin = None
        t_1: f32[4, 3] = torch.ops.aten.t.default(copy);  copy = None
        mm_1: f32[2, 3] = torch.ops.aten.mm.default(tangents_1, t_1);  tangents_1 = t_1 = None
        return pytree.tree_unflatten([mm, mm_1, None], self._out_spec)
```

I don't want the dependence on primals_2.

This example modeled based off of FSDP.

cc @bdhirsh @msaroufim @wconstab @anijain2305 @zou3519 @ngimel 

### Versions

main

**Comment by bdhirsh** [comment]
hmm, is this something that you think should be fixed in functionalization somewhere? This feels to me like an optimization: functionalization is doing as asked (functionalizing the `x.copy_(y)`), and now we need to claw back our memory usage somehow by taking advantage of the fact that `x`'s data doesn't actually matter anywhere after the last usage of `x`, but before the copy (similar to reinplacing).

**Comment by ezyang** [comment]
Yeah I would be fine with this being a separate optimization, eg decomp

**Comment by ezyang** [comment]
The important thing is it needs to happen before partitioning

**Comment by bdhirsh** [comment]
@voznesenskym just checking, but now that we have the set_() no-op logic, do you think this is still important / hi-pri for making tracing-fsdp performant?

I think it's still an issue either way, but how soon we should do it probably depends on whether it still impacts fsdp

**Comment by ezyang** [comment]
The old script has an error. New script

```
import torch
from torch._functorch.aot_autograd import aot_export_joint_simple
    
def f(x, weight):
    def cb(grad_out):
        y = grad_out.sin()  # stand-in for collective result
        weight.data.copy_(y)
        # Better would be:
        #   weight.data = y
        return grad_out
    r = x @ weight
    r.register_hook(cb) 
    return (r,)

gm = aot_export_joint_simple(
    f,
    [torch.randn(2, 2, requires_grad=True), torch.randn(2, 2)],
    trace_joint=True,
)
gm.print_readable()
```

This prints

```
class joint_helper(torch.nn.Module):
    def forward(self, primals, tangents):
        primals_1: "f32[2, 2]"; primals_2: "f32[2, 2]"; tangents_1: "f32[2, 2]"; 
    
        primals_1, primals_2, tangents_1, = fx_pytree.tree_flatten_spec([primals, tangents], self._in_spec)
        # No stacktrace found for following nodes
        mm: "f32[2, 2]" = torch.ops.aten.mm.default(primals_1, primals_2);  primals_1 = None
        sin: "f32[2, 2]" = torch.ops.aten.sin.default(tangents_1)
        detach: "f32[2, 2]" = torch.ops.aten.detach.default(primals_2)
        detach_1: "f32[2, 2]" = torch.ops.aten.detach.default(detach);  detach = None
        copy: "f32[2, 2]" = torch.ops.aten.copy.default(detach_1, sin);  detach_1 = sin = None
        t_1: "f32[2, 2]" = torch.ops.aten.t.default(copy)
        mm_1: "f32[2, 2]" = torch.ops.aten.mm.default(tangents_1, t_1);  tangents_1 = t_1 = None
        copy_: "f32[2, 2]" = torch.ops.aten.copy_.default(primals_2, copy);  primals_2 = copy = None
        return pytree.tree_unflatten([mm, mm_1, None], self._out_spec)
```

This seems... fine.
